stonesoup/reader/video.py

```diff
@@ -0,0 +1,218 @@
+# -*- coding: utf-8 -*-
+"""Video readers for Stone Soup.
+
+This is a collection of video readers for Stone Soup, allowing quick reading
+of video data/streams.
+"""
+
+from abc import abstractmethod
+import datetime
+import numpy as np
+import ffmpeg
+import moviepy.editor as mpy
+import threading
+from queue import Queue
+
+from ..base import Property
+from ..buffered_generator import BufferedGenerator
+from ..types.sensordata import ImageFrame
+from .base import SensorDataReader
+from .file import FileReader
+from .url import UrlReader
+
+
+class FrameReader(SensorDataReader):
+    """FrameReader base class
+
+    A FrameReader produces :class:`~.SensorData` in the form of
+    :class:`~ImageFrame` objects.
+    """
+
+    @property
+    def frame(self):
+        return self.sensor_data
+
+    @abstractmethod
+    @BufferedGenerator.generator_method
+    def frames_gen(self):
+        """Returns a generator of frames for each time step.
+
+        Yields
+        ------
+        : :class:`datetime.datetime`
+            Datetime of current time step
+        : set of :class:`~.ImageFrame`
+            Generated frame in the time step
+        """
+        raise NotImplementedError
+
+    @BufferedGenerator.generator_method
+    def sensor_data_gen(self):
```

Comment:
I am not too sure if this is correct/desired. 

My reasoning is that `FrameReader` objects generate frames (i.e. `ImageFrames`), rather than generic sensor data (i.e. `SensorData`). Therefore it made sense to have a `frame` property and `frames_gen()` method, in the same way `DetectionReaders` have `detections` and `detections_gen()`. 

Based on the above definitions, readers derived from `FrameReader` would only need to defined the `frame_gen()` method (as is done in the newly introduced `VideoClipReader` and `FFmpegVideoStreamReader` classes).

Reply:
This makes sense, as sensor data is more unique compared to other elements of Stone Soup, as you need a matching detector to make "sense" of it.